File: consensus-sim-checker/environment/reference/state_checker.zig
// Extracted from TigerBeetle src/testing/cluster/state_checker.zig
// https://github.com/tigerbeetle/tigerbeetle
//
// The state checker validates strict serializability by maintaining a canonical
// commit history and comparing each replica's state against it.
//
// Key data structures:
//   commits: ArrayList — canonical commit history (one entry per committed op)
//   commit_mins: [members_max]u64 — latest op committed by each replica
//
// Key invariant enforced in check_state():

pub fn check_state(state_checker: *StateChecker, replica_index: u8) !void {
    // ... (sync handling omitted) ...

    const commit_a = state_checker.commit_mins[replica_index];
    const commit_b = replica.commit_min;

    const checksum_a = state_checker.commits.items[commit_a].header.checksum;
    const checksum_b = /* replica's current checksum for commit_b */;

    // CRITICAL INVARIANT: if the op numbers match, checksums MUST match.
    // This detects divergent state even when a replica "recommits" the same op
    // (e.g. after a restart/replay). A replica that reports the same op number
    // but a different checksum indicates a consensus safety violation.
    assert((commit_a == commit_b) == (checksum_a == checksum_b));

    if (checksum_a == checksum_b) return;

    // Replica has advanced: commit_b should be exactly commit_a + 1
    assert(commit_b < commit_a or commit_a + 1 == commit_b);
    state_checker.commit_mins[replica_index] = commit_b;

    // Verify the new commit matches canonical history
    if (replica.commit_min < state_checker.commits.items.len) {
        const commit = &state_checker.commits.items[commit_b];
        assert(checksum_b == commit.header.checksum);
        commit.replicas.set(replica_index);
        return;
    }

    // New commit extends canonical history
    assert(header_b.?.parent == checksum_a);
    assert(header_b.?.op > 0);
    // ... (append to commits) ...
}

File: consensus-sim-checker/solution/state_checker.py
"""Consensus safety checker for strict serializability."""


class SafetyViolation(Exception):
    pass


class StateChecker:
    def __init__(self, replica_count: int):
        self.replica_count = replica_count
        self.commits = {
            0: {
                'checksum': 0,
                'parent_checksum': 0,
                'replicas': set(range(replica_count)),
            }
        }
        self.commit_mins = [0] * replica_count

    def on_commit(self, replica_index: int, op: int, checksum: int,
                  parent_checksum: int) -> None:
        assert 0 <= replica_index < self.replica_count
        assert op > 0

        current_min = self.commit_mins[replica_index]

        if op <= current_min:
            if op in self.commits and self.commits[op]['checksum'] != checksum:
                raise SafetyViolation(
                    f"Replica {replica_index} recommitted op {op} with "
                    f"divergent checksum: expected {self.commits[op]['checksum']}, "
                    f"got {checksum}"
                )
            return

        if op != current_min + 1:
            raise SafetyViolation(
                f"Replica {replica_index} skipped op: "
                f"expected {current_min + 1}, got {op}"
            )

        if current_min in self.commits:
            expected_parent = self.commits[current_min]['checksum']
            if parent_checksum != expected_parent:
                raise SafetyViolation(
                    f"Replica {replica_index} has wrong parent for op {op}: "
                    f"expected {expected_parent}, got {parent_checksum}"
                )

        if op in self.commits:
            if self.commits[op]['checksum'] != checksum:
                raise SafetyViolation(
                    f"Replica {replica_index} diverged at op {op}: "
                    f"existing {self.commits[op]['checksum']}, "
                    f"new {checksum}"
                )
            self.commits[op]['replicas'].add(replica_index)
        else:
            self.commits[op] = {
                'checksum': checksum,
                'parent_checksum': parent_checksum,
                'replicas': {replica_index},
            }

        self.commit_mins[replica_index] = op

    def check_convergence(self, replica_indices: set) -> bool:
        if not self.commits:
            return True
        max_op = max(self.commits.keys())
        return all(self.commit_mins[r] == max_op for r in replica_indices)

    def get_commit_count(self) -> int:
        return len(self.commits)

    def get_replica_commit_min(self, replica_index: int) -> int:
        return self.commit_mins[replica_index]
